### PR TITLE
#42316 Wait for `run` goroutine to exit before `Close`

### LINKED ISSUE #42316: Possible race condition between RingLogger's Close() and run()
Possible race condition between RingLogger's Close() and run()

We are running Docker on AWS ECS with the fluentd log driver configured, and periodically notice the following panic:

```panic: send on closed channelgoroutine 840293412 [running]:
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).appendBuffer(0xc0043a11e0, 0xc0
084da1e0, 0xc0084da180, 0xc0084da1b0)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:306 +0x4c
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).postRawData(0xc0043a11e0, 0xc00
84da1e0, 0x1d, 0x1668656)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:220 +0x3d
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).EncodeAndPostData(0xc0043a11e0,
 0xc00ab8fa80, 0x1d, 0x1668656, 0xed814becb, 0x0, 0x560000475460, 0xc0084da180, 0x98, 0xc003067a68)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:210 +0x159
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).PostWithTime(0xc0043a11e0, 0xc0
0ab8fa80, 0x1d, 0x1668656, 0xed814becb, 0x0, 0x56000047ab60, 0xc0084da150, 0x0, 0xc001770a80)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:201 +0x753
github.com/docker/docker/daemon/logger/fluentd.(*fluentd).Log(0xc008573cc0, 0xc0001e4c40, 0x0, 0x0)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/fluentd/fluentd.go
:177 +0x3cb
github.com/docker/docker/daemon/logger.(*RingLogger).run(0xc00d8e2a50)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/ring.go:119 +0x7d
created by github.com/docker/docker/daemon/logger.newRingLogger
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/ring.go:43 +0x176
```
        
The panic arises from [`fluent-logger-golang`](https://github.com/fluent/fluent-logger-golang/blob/3fe534f4a874ee46d3fe6982f0161194179403c3/fluent/fluent.go#L342) when there is an attempt to log something after the logger has been closed.

This seems to be possible if the RingBuffer gets closed between [ring.go:125](https://github.com/moby/moby/blob/e0170da0dc6e660594f98bc66e7a98ce9c2abb46/daemon/logger/ring.go#L125) grabbing a message from the buffer and pushing it to the underlying logger a few lines later.

**Steps to reproduce the issue:**
don't have an easy reproduction, but we're getting about 4 instances of this per day in a moderately sized ECS fleet (50-100 EC2 instances running several hundred containers)

**Describe the results you received:**
see panic stacktrace above

**Describe the results you expected:**
smooth operation

**Additional information you deem important (e.g. issue happens only occasionally):**
n/a

**Output of `docker version`:**

```
Client:
 Version:           19.03.6-ce
 API version:       1.40
 Go version:        go1.13.4
 Git commit:        369ce74
 Built:             Fri May 29 04:00:45 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.6-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       369ce74
  Built:            Fri May 29 04:02:31 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.2
  GitCommit:        ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 19.03.6-ce
 Storage Driver: devicemapper
  Pool Name: docker-docker--pool
  Pool Blocksize: 524.3kB
  Base Device Size: 10.74GB
  Backing Filesystem: ext4
  Udev Sync Supported: true
  Data Space Used: 744MB
  Data Space Total: 106.3GB
  Data Space Available: 105.6GB
  Metadata Space Used: 3.084MB
  Metadata Space Total: 25.17MB
  Metadata Space Available: 22.08MB
  Thin Pool Minimum Free Space: 10.63GB
  Deferred Removal Enabled: true
  Deferred Deletion Enabled: true
  Deferred Deleted Device Count: 0
  Library Version: 1.02.135-RHEL7 (2016-11-16)
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.14.133-88.112.amzn1.x86_64
 Operating System: Amazon Linux AMI 2018.03
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 30.41GiB
 Name: ip-10-7-43-255
 ID: EG5L:TJJG:VR2H:J7CP:WZ76:P2QS:PGWN:WEYP:SIRJ:A7ZM:5WIY:QSNN
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 51
  Goroutines: 74
  System Time: 2021-04-24T00:13:59.352411722Z
  EventsListeners: 1
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: the devicemapper storage-driver is deprecated, and will be removed in a future release.
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS ECS, with unmodified ECS-optimized Amazon Linux AMI (_not Amazon Linux 2_)

### PR DESCRIPTION
**Pull request by anujva**
Fixes https://github.com/moby/moby/issues/42316

  ###
The underlying Loggers Close(), function can be called with the the
run() goroutine still writing to the driver.
This is causing the fluentd-golang-logger to panic cause it doesn't
defensively check for the closing of the channel before writing to it.
It relies on the docker daemon to keep the contract of not calling Log()
if Close() has already been called.

**- What I did**
Synchronize the close of the underlying logger only after the run goroutine has exited.

**- How I did it**
Added a sync.Waitgroup to be instantiated withe the RingLogger and then added a wait condition of the goroutine to exit,
before Close is called on the underlying logger.

**- How to verify it**
I am not sure how best to verify it. Let me know if I should add a unit test for this change or something. 

**- Description for the changelog**

Fixes the race condition between RingLogger Close() and run() goroutine.

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

**Comment by anujva** [comment]
#42316 should be fixed by this. This panic is being caused by the fluentd-golang-logger not being defensive about the close of its channel. A comparable implementation of splunk actually uses a RWMutex to solve this problem, But this fix would solve the problem for both underlying drivers not needing any synchronization primitives.

File: daemon/logger/ring.go [hunk 1]
```diff
@@ -111,13 +114,15 @@ func (r *RingLogger) Close() error {
 			logErr = true
 		}
 	}
+	r.wg.Wait()
```
Comment (cpuguy83) [inline daemon/logger/ring.go:117]: Maybe this should go after setClosed?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: daemon/logger/ring.go (hunk 1 above)
Comment (anujva) [reply daemon/logger/ring.go:117]: It works either ways.. buffer.Close() has the right semantics to not have a race condition. I will move it up if that is more preferred.

Comment (cpuguy83) [reply daemon/logger/ring.go:117]: My thought here is to not have two things attempting to log at the same time.

Comment (anujva) [reply daemon/logger/ring.go]: Sounds good.. I moved it after the `setClosed()` call.

File: daemon/logger/ring.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ type RingLogger struct {
 	l         Logger
 	logInfo   Info
 	closeFlag int32
+	wg        *sync.WaitGroup
```
Comment (cpuguy83) [inline daemon/logger/ring.go:22]: Sorry I didn't catch this earlier, `sync.WaitGroup` should be used as a value here.

Then at line 45 there is no need to initialize it either.

Comment (cpuguy83) [reply daemon/logger/ring.go:22]: That is, not a pointer.

Comment (anujva) [reply daemon/logger/ring.go]: Updated. Please check. 

FWIW, the sync.Waitgroup funcs `Add` and `Wait` [0] are on the pointer receiver, so I think both would have worked. Curious why you preferred to use it as a value.

[0] https://golang.org/pkg/sync/#WaitGroup.Add

Comment (cpuguy83) [reply daemon/logger/ring.go]: It's just the "normal" way to use it... or at least how we use it throughout the codebase.
Only take the pointer when you need to pass it to another function.

Comment (anujva) [reply daemon/logger/ring.go]: Okay.. thank you for explaining. I have updated it.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**